File: src/frontend/src/app/models/subcategory.ts
import {Category} from "./category";
import {Item} from "./item";

export class Subcategory {
    private _id: number = 0;
    private _name: string = "";
    private _category: Category;
    private _items: Item[] = [];

    constructor(name: string, category: Category) {
        this._name = name;
        this._category = category;
    }

    get id(): number {
        return this._id;
    }

    set id(value: number) {
        this._id = value;
    }

    get name(): string {
        return this._name;
    }

    set name(value: string) {
        this._name = value;
    }

    get category(): Category | undefined {
        return this._category;
    }

    set category(value: Category | undefined) {
        this._category = <Category>value;
    }

    get items(): Item[] {
        return this._items;
    }

    set items(value: Item[]) {
        this._items = value;
    }
}

File: src/backend/java/org/example/Angular/entities/Subcategory.java
package org.example.Angular.entities;

import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;

import javax.persistence.*;
import java.util.HashSet;
import java.util.Set;

@Entity
@JsonIdentityInfo(
        property = "id",
        generator = ObjectIdGenerators.PropertyGenerator.class
)
public class Subcategory {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int id;

    private String name;
    private String image;

    private int category;

    @OneToMany(fetch = FetchType.EAGER)
    @JoinTable(name = "subcategory_item",
            joinColumns = @JoinColumn(name = "subcategory_id"),
            inverseJoinColumns = @JoinColumn(name = "item_id"))
    private Set<Item> items = new HashSet<>();

//  Constructors

    public Subcategory() {
    }

    public Subcategory(String name) {
        this.name = name;
    }

    public Subcategory(String name, int category) {
        this.name = name;
        this.category = category;
    }

    //  Getters & Setters

    public int getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getImage() {
        return image;
    }

    public void setImage(String image) {
        this.image = image;
    }

    public int getCategory() {
        return category;
    }

    public void setCategory(int category) {
        this.category = category;
    }

    public Set<Item> getItems() {
        return items;
    }

    public void addItem(Item item) {
        items.add(item);
    }

    @Override
    public String toString() {
        return "Subcategory{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", image='" + image + '\'' +
                ", category=" + category +
                ", items=" + items +
                '}';
    }
}